Cics web command fix v2

mm-broadcom opened the pull request:
Implementation of IBM's grammar for the WEB CICS command such that fulfills all of those requirements but is flexible enough for what the compiler accepts.

[Here is the V1 Branch](https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/compare/development...mm-broadcom:che-che4z-lsp-for-cobol:CICS-WEB-Command-Fix) for reference. (Due to the amount of time between creation, merge/updates from Development and such, it was easier to make a new branch and port those over rather than shoehorn it in.)

## How Has This Been Tested?

Wrote tests with example calls based upon IBM's documentation. 

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Copy changes from V1 branch

Copied the changes from the V1 branch to this one. (Too many commits and conflicts to rebase easily.)

mm-broadcom committed:
Test fixes

mm-broadcom committed:
WEB Rule rewrites

mm-broadcom committed:
Fix existing test, made rules a bit more flexible

Fixed an existing test as it was using a token/command CHARACTERSET which did not exist in the CICS Receive rule for Clients. Added some more flexibility to the new rules as well.

mm-broadcom committed:
Initial outline of checks for WEB's commands

Initial outline of checks for WEB's commands. Committing to save changes so far.

mm-broadcom committed:
Mutually Exclusive function, expanded parameters

Added a function to check for mutually exclusive rules.

mm-broadcom committed:
More utility functions/overloads

More utility functions and overloads.

mm-broadcom committed:
WEB rule checking and fixes

Added more WEB command rule verification, and fixes/consolidation of rules as well to correspond with the new checks.

mm-broadcom committed:
Allow cics_handle_response mixing

Allow cics_handle_response to be mixed in with each of the WEB command subtypes.

mm-broadcom committed:
Provide first word if prior rule seen

Only get the first word if prior rule seen. Avoids a lot of overhead in the case of a ParserRuleContext.

mm-broadcom committed:
Have error message only list the options

Due to how expansive and computationally expensive the .getText() call on a Parser Rule Context could be, it was removed. The list of options along with the context thereof _should_ provide enough information as to where the error is.

mm-broadcom committed:
Merge branch 'development' into CICS-WEB-Command-Fix-V2

mm-broadcom committed:
Change if-else cascade into Switch statement

Change the multitude of if-else statements into a switch statement using the rule's index as the basis to go off of. This uses a primitive type rather than blindly asking every possible class if it is it.

mm-broadcom committed:
Added complete duplicate checking

Added complete duplicate checking for WEB commands.

mm-broadcom committed:
Merge remote-tracking branch 'upstream/development' into CICS-WEB-Command-Fix-V2

mm-broadcom committed:
Change to <Integer, ErrorSeverity> vs <String, ErrorSeverity>

Change to <Integer, ErrorSeverity> vs <String, ErrorSeverity>

mm-broadcom committed:
Merge remote-tracking branch 'upstream/development' into CICS-WEB-Command-Fix-V2

mm-broadcom committed:
Fixed checkHasRequired functions

Fixed checkHasRequired functions and overloads.

mm-broadcom committed:
Add WEB to optionsMap

mm-broadcom committed:
Iterate over subrules

mm-broadcom committed:
Fix/expand tests

mm-broadcom committed:
Fixed tests

mm-broadcom committed:
Removed unused function

mm-broadcom committed:
Move invalid test case

mm-broadcom commented:
@slavek-kucera Would it be possible for you to review this?

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -789,82 +789,96 @@ cics_waitcics: WAITCICS (ECBLIST cics_value | NUMEVENTS cics_data_value | PURGEA
-cics_web_write: WRITE (HTTPHEADER cics_data_area | NAMELENGTH cics_data_value | SESSTOKEN cics_data_value |
-                VALUE cics_data_area | VALUELENGTH cics_data_value | cics_handle_response)+;
+
+cics_web_close: CLOSE SESSTOKEN cics_data_value;
```
`cics_handle_response` seems to be a valid option.

mm-broadcom replied:
Resolved in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -111,14 +111,70 @@ protected boolean checkHasMandatoryOptions(List<?> rules, ParserRuleContext ctx,
   }
 
+  protected void checkHasMandatoryOptions(TerminalNode rules, ParserRuleContext ctx, String options) {
+    ArrayList<TerminalNode> tempList = new ArrayList<>();
+    tempList.add(rules);
+    checkHasMandatoryOptions(tempList, ctx, options);
```
This looks equivalent to just an empty function.

mm-broadcom replied:
Replaced the one usage of this in the latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -127,6 +183,55 @@ protected void checkHasIllegalOptions(List<?> rules, String options) {
+   * @param options Options checked to insert into error message
+   */
+  protected void checkHasIllegalOptions(TerminalNode rule, String options) {
+    if (!rule.getText().isEmpty()) {
```
This condition is just always true, isn't it?

mm-broadcom replied:
No, it can be of the type but not have any content/text associated with it. You see this with the `*` rules quite often.

slavek-kucera replied:
But this is a TerminalNode - it is created to represent a matched token.

mm-broadcom replied:
This function has been removed in the current commit. It was a lesser form of the checkPrerequisiteIsMet function.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -127,6 +183,55 @@ protected void checkHasIllegalOptions(List<?> rules, String options) {
+      }
+
+      if (ParserRuleContext.class.isAssignableFrom(rule.getClass())) {
+        if (!((ParserRuleContext) rule).isEmpty()) {
```
I think this is also always true.

mm-broadcom replied:
Not always, we care if there is content not that the current element in the list/collection is of a given type.

slavek-kucera replied:
I don't think that the `isEmpty` does what you think.
https://github.com/antlr/antlr4/blob/933c4e0a0d4d5accfc2f9fa416e2c0558b4512e0/runtime/Java/src/org/antlr/v4/runtime/RuleContext.java#L100

mm-broadcom replied:
That is exactly _why_ I call isEmpty as it checks to see if a given RuleContext or derivative thereof has been called within the given instance of the rule.

The purpose of this function is to take a collection of _possible_ tokens and/or rules and see if more than one has been called. If so, insert an exception/warning message into the LSP at the given context to inform the user accordingly.

slavek-kucera replied:
- this expression will be always true
- neither of the top level conditions are true - `rules` is a generally "list of lists".

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -127,6 +183,55 @@ protected void checkHasIllegalOptions(List<?> rules, String options) {
+          rulesSeen++;
+        }
+      } else if (TerminalNode.class.isAssignableFrom(rule.getClass())) {
+        if (!((TerminalNode) rule).getText().isEmpty()) {
```
always true

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -127,6 +183,55 @@ protected void checkHasIllegalOptions(List<?> rules, String options) {
+          continue;
+      }
+
+      if (ParserRuleContext.class.isAssignableFrom(rule.getClass())) {
```
Is this the same as `rule instanceof ParserRuleContext`?

mm-broadcom replied:
No, `instanceof` requires the type to be known at compile time not in a dynamic capacity. As we're checking whether it is a ParserRuleContext _or_ a TerminaNode, we cannot really use instanceof.

slavek-kucera replied:
Not that this issue matters much, but ParserRuleContext and TerminalNode are known at compile time.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java:
```diff
@@ -111,14 +111,70 @@ protected boolean checkHasMandatoryOptions(List<?> rules, ParserRuleContext ctx,
+   * @param ctx - The overall context.
+   * @param options - String of the element that is required.
+   */
+  protected <E extends ParseTree> void checkHasRequiredOption(List<E> requiredContext, List<E> optionalContext, ParserRuleContext ctx, String options) {
```
Could you come up with a better name for this function. It currently sounds synonymous to `checkHasMandatoryOptions`.

mm-broadcom replied:
Renamed in latest commit.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSWebOptionsCheckUtility.java:
```diff
@@ -0,0 +1,507 @@
+    private void checkClose(CICSParser.Cics_web_closeContext ctx) {
+        List<TerminalNode> ruleList = new ArrayList<>();
+        ruleList.add(ctx.SESSTOKEN()); // SESSTOKEN in CLOSE does not utilize the repeat notation.
+        checkHasMandatoryOptions(ruleList, ctx, "SESSTOKEN");
```
I'd expect just
```suggestion
        checkHasMandatoryOptions(ctx.SESSTOKEN(), ctx, "SESSTOKEN");
```

mm-broadcom replied:
Fixed in latest commit.

slavek-kucera commented on server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCicsWebStatement.java:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    @Test
+    void testReadInvalid() {
```
Add a few more tests to covert the error cases.

mm-broadcom committed:
Add cics_handle_response to WEB CLOSE command.

mm-broadcom committed:
Change function name to lessen confusion.

mm-broadcom committed:
Streamlined checkClose

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4:
```diff
@@ -789,82 +789,96 @@ cics_waitcics: WAITCICS (ECBLIST cics_value | NUMEVENTS cics_data_value | PURGEA
-cics_web_write: WRITE (HTTPHEADER cics_data_area | NAMELENGTH cics_data_value | SESSTOKEN cics_data_value |
-                VALUE cics_data_area | VALUELENGTH cics_data_value | cics_handle_response)+;
+
+cics_web_close: CLOSE SESSTOKEN cics_data_value cics_handle_response+;
```
Both orderings are allowed.

mm-broadcom replied:
Fixed in latest commit.

mm-broadcom committed:
Remove checkHasIllegalOptions

Remove that function as checkPrerequisiteIsMet covers the same purpose with less overhead from the developer.

mm-broadcom committed:
Update WEB CLOSE to be flexible on ordering.

mm-broadcom committed:
Update checkMutuallyExclusiveOptions

mm-broadcom committed:
Remove extra checkHasMandatoryOptions function

mm-broadcom committed:
Fixed misspelling.

mm-broadcom committed:
Merge remote-tracking branch 'upstream/development' into CICS-WEB-Command-Fix-V2

mm-broadcom committed:
Duplicate rule check

mm-broadcom committed:
Fix error/test message

mm-broadcom committed:
Add LENGTH to web_statuscode

mm-broadcom committed:
Fix spelling error in test variable name, add invalid test

mm-broadcom committed:
Move send server body to their main rules, fixed rule index

mm-broadcom committed:
Add check for lists in checkMutuallyExclusiveOptions

Add check for lists in checkMutuallyExclusiveOptions, only count them once per iteration. Set a boolean to denote if the current "rule" being checked is a list and if so, grab the first relevant entry to use for retrieving the locality.

mm-broadcom committed:
Add test for WEB SEND (server)

mm-broadcom committed:
Fixed checkstyle error

mm-broadcom committed:
Add check for empty list

mm-broadcom committed:
Fixed List cast to remove the suppressed warning.

mm-broadcom committed:
Fixed checks/tests

mm-broadcom committed:
Added another purposeful invalid test case

mm-broadcom committed:
Remove variable that was used for debugging

Remove variable that was used for debugging purposes. (Gave an easy way to see the incoming ruleindex which allowed for finding that the WEB SEND rule index's case was accidentally using the base SEND command one.)
